simplify steadystate logic

FFroehlich opened the pull request.

paulstapor commented:
That one was on my to-do-list anyway... Would have been good to discuss this prior to you putting effort in it...
The logic here isn't simple, but that's not (let's say, maybe not only) because it's not perfectly programmed, but rather because the problem as such is intricate. You probably saw it from the fact that you want to simplify the code, but actually added more lines than you removed...

So, give me some background please, as the PR has no description:
 - What is precisely you aim of the code? What does the current code not do? What should the new code do (better)?
 - (How) Does this relate to #1073 ?

FFroehlich commented:
> That one was on my to-do-list anyway... Would have been good to discuss this prior to you putting effort in it...
> The logic here isn't simple, but that's not (let's say, maybe not only) because it's not perfectly programmed, but rather because the problem as such is intricate. You probably saw it from the fact that you want to simplify the code, but actually added more lines than you removed...

The big issue is that there was a single function that served 3 different purposes (check whether simulation needs to be stored, whether simulation needs to be run with sensitivities, and whether sensitivities are to be computed using the implicit function theorem). This is in violation of the single-responsibility principle (https://en.wikipedia.org/wiki/Single-responsibility_principle), giving rise to code execution that is so complicated that eve you got confused about it `https://github.com/AMICI-dev/AMICI/pull/1461`

> 
> So, give me some background please, as the PR has no description:
> 
> * What is precisely you aim of the code? What does the current code not do? What should the new code do (better)?

1) the `simulationFSA` sensitivity mode is intended to be used when the Jacobian is singular. In that case, sensitivity computation via the implicit function theorem should never be used. 
2) Having a mode where you use forward sensitivities only when simulation is run is not practically relevant since I cannot tell any situations when the user would want to do that. Moreover, it is questionable whether forward sensitivities will be accurate when newtons method was used prior to simulation since the sensitivity initialization are not updated by newtons method. On the one hand, using newtons method adds little benefit since simulation has to be run until convergence of sensitivities. While you may get lucky and they may converge eventually, you are in big trouble for multistable systems, but whether this will work in multistable systems is unclear.
3) You cannot run adjoint sensitivity analysis and then use the same solver for postequilibration using sensitivity analysis. I added a check that detects that situation and raises an issue
4) Many of the test that were supposed to test combinations of forward and adjoint methods simply used newtons method to compute sensitivity and hence didn't really test what they were supposed to test. I added checks for preeq and posteq status flags that verifies that the right method were run.
5) Convergence of steadystate sensitivities were based on norm of x_, not sx_, which honestly just doesn't make too much sense.

> * (How) Does this relate to #1073 ?

It partially addresses the item `Add flag "simulationOnly" to steadyStateSensitivityMethod, which does not even try to launch the Newton solver.`

paulstapor commented on include/amici/steadystateproblem.h:
```diff
@@ -128,17 +128,45 @@ class SteadystateProblem {
+    bool checkNewtonAllowed(const Model &model, const Solver &solver);
+    
+    /**
+     * @brief Checks whether the simulation solver or initialization provided sensitivities are correct
```
I don't think the docstring is correct... This method actually checks whether state sensitivities have already been computed.

```suggestion
     * @brief Checks whether sensitivities have already been computed by the simulation solver 
```

FFroehlich replied:
> I don't think the docstring is correct... This method actually checks whether state sensitivities have already been computed.

I don't understand what makes you think the description is not correct. This also checks whether you can use `sx` from initialization.

paulstapor replied:
My understanding of "checking whether sensitivities are correct" is a finite difference check, which compares numerically approximated sensitivities to analytically computed ones. At least this is my very first thought, although it doesn't make sense in this context. Therefore, I'd suggest another docstring.

FFroehlich replied:
ah!

paulstapor reviewed (approved):
That's somewhat orthogonal to what I have planned to do with the steady state solver, therefore I'm not completely happy about it
If I'm not mistaken, the SteadyStateContext flag isn't used any more by now. In this case, please remove it also from include/amici/defines.h

FFroehlich commented on src/steadystateproblem.cpp:
```diff
@@ -26,19 +26,30 @@ SteadystateProblem::SteadystateProblem(const Solver &solver, const Model &model)
               throw AmiException("Preequilibration using adjoint sensitivities "
                                  "is not compatible with using forward "
                                  "sensitivities during simulation");
+          
+          if (solver.getSensitivityMethodPreequilibration()
+                  == SensitivityMethod::adjoint &&
+              model.getSteadyStateSensitivityMode() ==
+                  SteadyStateSensitivityMode::simulationFSA)
+              throw AmiException("Preequilibration using adjoint sensitivities "
+                                 "is not compatible with simulationFSA "
+                                 "sensitivity mode.");
```
> I understand that setting `SteadyStateSensitivityMode::simulationFSA` and allowing to use adjoint preequilibration is counter-intuitive. However, it would make much sense to allow the user to switch off the Newton solver and using adjoint preequilibration... Would be happy if there was a way to encode this...

How about setting the allowed newton steps to 0?

paulstapor replied:
Ideally, there would be the options newtonOnly, simulationFSA, and simulationASA and something that first tries Newton and uses simulation only if Newton fails... But no need to implement this now.

Something else: I'd advocate to move the steadyStateSensitivitiyMode from model to solver. The value may be model dpeendent, but it encodes the behavior of the solver. Therefore, I'd be happy if we could move this at some point. What's your opinion on this?

FFroehlich replied:
> Ideally, there would be the options newtonOnly, simulationFSA, and simulationASA and something that first tries Newton and uses simulation only if Newton fails... But no need to implement this now.

I think most of these settings can already be achieved with existing options.

> Something else: I'd advocate to move the steadyStateSensitivitiyMode from model to solver. The value may be model dpeendent, but it encodes the behavior of the solver. Therefore, I'd be happy if we could move this at some point. What's your opinion on this?

Yes, I agree, I found it confusing that this is in model, not solver.

FFroehlich commented on src/steadystateproblem.cpp:
```diff
@@ -53,6 +64,18 @@ void SteadystateProblem::workSteadyStateProblem(Solver *solver, Model *model,
     } else {
+        /* don't use checkUseFSASensis here, since steady_state_status_ checks
+           are meaningless at this point */
+        if (solver->getSensitivityOrder() >= SensitivityOrder::first &&
+            model->getSteadyStateSensitivityMode() ==
+                SteadyStateSensitivityMode::simulationFSA &&
+            solver->getSensitivityMethod() != SensitivityMethod::forward) {
```
This check may have become obsolete after adding the respective checks to the constructor.

FFroehlich commented:
> That's somewhat orthogonal to what I have planned to do with the steady state solver, therefore I'm not completely happy about it

If you plan stuff for a year or so and never follow through, you kinda have to live with the possibility that someone starts working on it beforehand.

> If I'm not mistaken, the SteadyStateContext flag isn't used any more by now. In this case, please remove it also from include/amici/defines.h
 
👍

paulstapor commented:
> > That's somewhat orthogonal to what I have planned to do with the steady state solver, therefore I'm not completely happy about it
> 
> If you plan stuff for a year or so and never follow through, you kinda have to live with the possibility that someone starts working on it beforehand.

No problem with somebody else working on it at all. But I created a milestone and organized issues for this for the precise reason that one can discuss the way how its implemented prior to it being done...

FFroehlich commented:
> That's somewhat orthogonal to what I have planned to do with the steady state solver, therefore I'm not completely happy about it

@paulstapor I added a mixed mode which restores the possibility to use combinations of FSA with newton methods

FFroehlich commented:
okay at this point I simply give up and consider the simulationFSA functionality broken. 

https://github.com/AMICI-dev/AMICI/blob/4e671a5be75e19da81b0946a3e16a913b7fe725b/python/tests/test_preequilibration.py#L307 never ran because once a generator is used, its empty.
